README
COM S 327
Assignment#1.10
Jooseung Song
rlg327.cpp

Added place_lava_water. characters lose 2 hp on lava and lose 2 speed on water.
PC regenerate 2 hp every 5 step. press 'c' to see description of anything in selected cell
Had a segmentation issue associated with io_nearest_visible_monster, but seems to be resolved.


\\\\\\

CHANGELOG
Jooseung Song, ID: 351320970
1/20/2018: Added isSafe and generateRoom function. Can now print layouts and place rooms.
1/21/2018: Fixed simple typo and reduced run time
1/22/2018: Tried to make simple pathfinding code. Function sort added, changed int array map to struct array.
1/23/2018: Added generateCorridor and draw function.
1/23/2018: Fixed segmentation fault error.
1/29/2018: Added dungeon structure. Deleted public map data. Added init_dungeon, load_dungeon, render_dungeon.
1/30/2018: Added save and load. Can execute both at a time. Seems fine when running with single source code.
1/30/2018: Fixed segmentation faults on running load function.
2/05/2018: Starting work based on professor Sheaffer's work which was distributed through canvas.
2/05/2018: Added PC structure, modified dungeon structure
2/10/2018: Added dijkstra_path and dijkstra_path_t function based on dijkstra_corridor function.
2/12/2018: Fixed dijkstra_path and dijkstra_path_t, added distance comparing functions.
2/12/2018: Modified place_room and fixed infinite loop.
2/12/2018: Modified render_dungeon function to add PC.
2/12/2018: Added place_PC function and modified main.
2/13/2018: Added render_distance_map and render_tunnel_distance_map.
2/18/2018: Starting work based on professor Sheaffer's work which was distributed through canvas.
2/18/2018: Fixed infinite loop.
2/19/2018: Modified render_dungeon.
2/19/2018: Added 8 move functions and modified main.
2/25/2018: Starting work based on last assignment
2/26/2018: Applied Ncurses modified most c files as followed.
3/3/2018: Deleted random PC movement and PC now cannot tunnel.
3/4/2018: Added key-map and modified main fuction. 
3/4/2018: Added menu function.
3/5/2018: Cannot get arrow-keys with getch()
3/6/2018: Finished working on menu.
3/16/2018: Starting work based on last assignment
3/19/2018: Changed most of C files to C++ except rgl327.c
3/19/2018: Fog of war working
3/20/2018: 'f' key and teleport mode working
3/20/2018: seg fault issue partially fixed
3/25/2018: Starting work based on last assignment
3/27/2018 Wrote all code to read and display everything from txt file
4/1/2018: Starting work based on last assignment
4/2/2018: Added object class
4/3/2018: Modified display function to add colors
4/3/2018: Modified can_see, gen_monster. 
4/3/2018: Added gen_object function from gen_monster.
4/9/2018: Starting work based on last assignment
4/9/2018: Added io_list_inventory, io_list_equuipment
4/9/2018: Modified description class
4/9/2018: Modified do_combat
4/10/2018: Added io_look class
4/10/2018: Added pc_addCarry, pc_dropItem, pc_wear, pc_takeOff, pc_inspect, pc_update
4/10/2018: Added all keys
4/14/2018: Starting work based on last assignment
4/15/2018: Added ter_water, ter_lava
4/15/2018: Added place_lava_water
4/16/2018: Modified move_character to make ter_water, ter_lava work
4/16/2018: Modified move_character and added regenerating function
4/16/2018: Added info in io.cpp
4/17/2018: Fixed segmentation issue associated with io_nearest_visible_monster
4/17/2018: Added 'c' key to work with info function

move.cpp
#include "move.h"

#include <unistd.h>
#include <stdlib.h>
#include <assert.h>

#include "dungeon.h"
#include "heap.h"
#include "move.h"
#include "npc.h"
#include "pc.h"
#include "character.h"
#include "utils.h"
#include "path.h"
#include "event.h"
#include "io.h"
#include "npc.h"

void do_combat(dungeon *d, character *atk, character *def)
{
  // int can_see_atk, can_see_def;
  const char *organs[] = {
    "liver",                   /*  0 */
    "pancreas",                /*  1 */
    "heart",                   /*  2 */
    "eye",                     /*  3 */
    "arm",                     /*  4 */
    "leg",                     /*  5 */
    "intestines",              /*  6 */
    "gall bladder",            /*  7 */
    "lungs",                   /*  8 */
    "hand",                    /*  9 */
    "foot",                    /* 10 */
    "spinal cord",             /* 11 */
    "pituitary gland",         /* 12 */
    "thyroid",                 /* 13 */
    "tongue",                  /* 14 */
    "bladder",                 /* 15 */
    "diaphram",                /* 16 */
    "stomach",                 /* 17 */
    "pharynx",                 /* 18 */
    "esophagus",               /* 19 */
    "trachea",                 /* 20 */
    "urethra",                 /* 21 */
    "spleen",                  /* 22 */
    "ganglia",                 /* 23 */
    "ear",                     /* 24 */
    "subcutaneous tissue"      /* 25 */
    "cerebellum",              /* 26 */ /* Brain parts begin here */
    "hippocampus",             /* 27 */
    "frontal lobe",            /* 28 */
    "brain",                   /* 29 */
  };
  int part;
  uint32_t damageDone = 0;
  uint16_t i;


  if (def->alive) {
    // def->alive = 0;
    // charpair(def->position) = NULL;
    
    if (atk == d->PC) {
      // d->num_monsters--;
      for(i=0; i < d->PC->equipSlots.size()-1; i++){
        if(d->PC->equipSlots[i] != NULL)
          damageDone += d->PC->equipSlots[i]->roll_dice();
      }
      damageDone += atk->damage->roll();

      if(damageDone > def->hp){
        //pc killed monster
        def->alive = 0;
        charpair(def->position) = NULL;
        d->num_monsters--;
        io_queue_message("%s smites the %s", atk->name, character_get_name(def));
        atk->kills[kill_direct]++;
        atk->kills[kill_avenged] += (def->kills[kill_direct] +
                                     def->kills[kill_avenged]);
      }else{
        def->hp -= damageDone;
        io_queue_message("Hit %s for %u damage", character_get_name(def), damageDone);
      }
    } else {
      damageDone += atk->damage->roll();

      if(damageDone > def->hp){
        def->alive = 0;
        charpair(def->position) = NULL;
        if ((part = rand() % (sizeof (organs) / sizeof (organs[0]))) < 26) {
          io_queue_message("As %s%s eats your %s,", is_unique(atk) ? "" : "the ",
                          atk->name, organs[rand() % (sizeof (organs) /
                                                      sizeof (organs[0]))]);
          io_queue_message("   ...you wonder if there is an afterlife.");
          /* Queue an empty message, otherwise the game will not pause for *
          * player to see above.                                          */
          io_queue_message("");
        } else {
          io_queue_message("Your last thoughts fade away as "
                          "%s%s eats your %s...",
                          is_unique(atk) ? "" : "the ",
                          atk->name, organs[part]);
          io_queue_message("");
        }
        atk->kills[kill_direct]++;
        atk->kills[kill_avenged] += (def->kills[kill_direct] +
                                      def->kills[kill_avenged]);
      }else{
        def->hp -= damageDone;
        io_queue_message("%s took %u damage", def->name, damageDone);
      }
      /* Queue an empty message, otherwise the game will not pause for *
       * player to see above.                                          */
      io_queue_message("");
    }
  }

  // if (atk == d->PC) {
  //   io_queue_message("You smite %s%s!", is_unique(def) ? "" : "the ", def->name);
  // }

  // can_see_atk = can_see(d, character_get_pos(d->PC),
  //                       character_get_pos(atk), 1, 0);
  // can_see_def = can_see(d, character_get_pos(d->PC),
  //                       character_get_pos(def), 1, 0);

  // if (atk != d->PC && def != d->PC) {
  //   if (can_see_atk && !can_see_def) {
  //     io_queue_message("%s%s callously murders some poor, "
  //                      "defenseless creature.",
  //                      is_unique(atk) ? "" : "The ", atk->name);
  //   }
  //   if (can_see_def && !can_see_atk) {
  //     io_queue_message("Something kills %s%s.",
  //                      is_unique(def) ? "" : "the helpless ", def->name);
  //   }
  //   if (can_see_atk && can_see_def) {
  //     io_queue_message("You watch in abject horror as %s%s "
  //                      "gruesomely murders %s%s!",
  //                      is_unique(atk) ? "" : "the ", atk->name,
  //                      is_unique(def) ? "" : "the ", def->name);
  //   }
  // }
}

void move_character(dungeon *d, character *c, pair_t next)
{
  if (charpair(next) &&
      ((next[dim_y] != c->position[dim_y]) ||
       (next[dim_x] != c->position[dim_x]))) {
    if(c == d->PC || charpair(next) == d->PC){
      do_combat(d, c, charpair(next));
    }
  } else {

    d->character_map[c->position[dim_y]][c->position[dim_x]] = NULL;
    c->position[dim_y] = next[dim_y];
    c->position[dim_x] = next[dim_x];
    d->character_map[c->position[dim_y]][c->position[dim_x]] = c;

    if(d->map[character_get_y(c)][character_get_x(c)] != ter_water && c->wet){
      c->speed += 2;
      c->wet = false;
    }else if(d->map[character_get_y(c)][character_get_x(c)] == ter_water && !c->wet){
      c->speed -= 2;
      if(c->speed < 1){
        c->speed = 1;
      }
      c->wet = true;
    }else if(d->map[character_get_y(c)][character_get_x(c)] == ter_lava){
      if(c == d->PC) {
        io_queue_message("it burns!");
      }
      if(2 > c->hp){
        if(c == d->PC){
          c->alive = 0;
          charpair(c->position) = NULL;
        }else{
          c->alive = 0;
          charpair(c->position) = NULL;
          d->num_monsters--;
        }
      }else{
        c->hp -= 2;
      }
    }
    if(c == d->PC){
      if(d->PC->regen == 5 && d->PC->hp != d->PC->maxHP){
        d->PC->hp += 2;
        d->PC->regen = 0;
        if(d->PC->hp > d->PC->maxHP){
          d->PC->hp = d->PC->maxHP;
        }
        io_queue_message("regenerated %s hp: %d", c->name, c->hp);
      }
      d->PC->regen ++;
    }
  }

  if (c == d->PC) {
    pc_reset_visibility((pc *) c);
    pc_observe_terrain((pc *) c, d);
  }
}

void do_moves(dungeon *d)
{
  pair_t next;
  character *c;
  event_t *e;

  /* Remove the PC when it is PC turn.  Replace on next call.  This allows *
   * use to completely uninit the heap when generating a new level without *
   * worrying about deleting the PC.                                       */

  if (pc_is_alive(d)) {
    /* The PC always goes first one a tie, so we don't use new_event().  *
     * We generate one manually so that we can set the PC sequence       *
     * number to zero.                                                   */
    e = (event_t *) malloc(sizeof (*e));
    e->type = event_character_turn;
    /* Hack: New dungeons are marked.  Unmark and ensure PC goes at d->time, *
     * otherwise, monsters get a turn before the PC.                         */
    if (d->is_new) {
      d->is_new = 0;
      e->time = d->time;
    } else {
      e->time = d->time + (1000 / d->PC->speed);
    }
    e->sequence = 0;
    e->c = d->PC;
    heap_insert(&d->events, e);
  }

  while (pc_is_alive(d) &&
         (e = (event_t *) heap_remove_min(&d->events)) &&
         ((e->type != event_character_turn) || (e->c != d->PC))) {
    d->time = e->time;
    if (e->type == event_character_turn) {
      c = e->c;
    }
    if (!c->alive) {
      if (d->character_map[c->position[dim_y]][c->position[dim_x]] == c) {
        d->character_map[c->position[dim_y]][c->position[dim_x]] = NULL;
      }
      if (c != d->PC) {
        event_delete(e);
      }
      continue;
    }

    npc_next_pos(d, (npc *) c, next);
    move_character(d, c, next);

    heap_insert(&d->events, update_event(d, e, 1000 / c->speed));
  }

  io_display(d);
  if (pc_is_alive(d) && e->c == d->PC) {
    c = e->c;
    d->time = e->time;
    /* Kind of kludgey, but because the PC is never in the queue when   *
     * we are outside of this function, the PC event has to get deleted *
     * and recreated every time we leave and re-enter this function.    */
    e->c = NULL;
    event_delete(e);
    io_handle_input(d);
  }
}

void dir_nearest_wall(dungeon *d, character *c, pair_t dir)
{
  dir[dim_x] = dir[dim_y] = 0;

  if (c->position[dim_x] != 1 && c->position[dim_x] != DUNGEON_X - 2) {
    dir[dim_x] = (c->position[dim_x] > DUNGEON_X - c->position[dim_x] ? 1 : -1);
  }
  if (c->position[dim_y] != 1 && c->position[dim_y] != DUNGEON_Y - 2) {
    dir[dim_y] = (c->position[dim_y] > DUNGEON_Y - c->position[dim_y] ? 1 : -1);
  }
}

uint32_t against_wall(dungeon *d, character *c)
{
  return ((mapxy(c->position[dim_x] - 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x] + 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] - 1) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] + 1) == ter_wall_immutable));
}

uint32_t in_corner(dungeon *d, character *c)
{
  uint32_t num_immutable;

  num_immutable = 0;

  num_immutable += (mapxy(c->position[dim_x] - 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x] + 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] - 1) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] + 1) == ter_wall_immutable);

  return num_immutable > 1;
}

static void new_dungeon_level(dungeon *d, uint32_t dir)
{
  /* Eventually up and down will be independantly meaningful. *
   * For now, simply generate a new dungeon.                  */

  switch (dir) {
  case '<':
  case '>':
    new_dungeon(d);
    break;
  default:
    break;
  }
}


uint32_t move_pc(dungeon *d, uint32_t dir)
{
  pair_t next;
  uint32_t was_stairs = 0;

  next[dim_y] = d->PC->position[dim_y];
  next[dim_x] = d->PC->position[dim_x];


  switch (dir) {
  case 1:
  case 2:
  case 3:
    next[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    next[dim_y]--;
    break;
  }
  switch (dir) {
  case 1:
  case 4:
  case 7:
    next[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    next[dim_x]++;
    break;
  case '<':
    if (mappair(d->PC->position) == ter_stairs_up) {
      was_stairs = 1;
      new_dungeon_level(d, '<');
    }
    break;
  case '>':
    if (mappair(d->PC->position) == ter_stairs_down) {
      was_stairs = 1;
      new_dungeon_level(d, '>');
    }
    break;
  }

  if (was_stairs) {
    return 0;
  }

  if ((dir != '>') && (dir != '<') && (mappair(next) >= ter_floor)) {
    move_character(d, d->PC, next);
    dijkstra(d);
    dijkstra_tunnel(d);

    return 0;
  }

  return 1;
}
